File: go/2_Add_Two_Numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    dummyHead := new(ListNode)
    curr := dummyHead
    carry := 0
    for l1 != nil || l2 != nil || carry != 0 {
        var x int
        var y int
        if l1 != nil {
            x = l1.Val
            l1 = l1.Next
        } 
        if l2 != nil {
            y = l2.Val
            l2 = l2.Next
        } 

        sum := carry + x + y
        carry = sum / 10

        curr.Next = &ListNode{sum % 10, nil}
        curr = curr.Next
    }

    return dummyHead.Next;
}

File: php/0000/2_Add_Two_Numbers.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution
{

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function addTwoNumbers($l1, $l2)
    {
        $dummyHead = new ListNode(0);
        $curr = $dummyHead;
        $carry = 0;
        while ($l1 !== null || $l2 !== null || $carry != 0) {
            $x = $l1 ? $l1->val : 0;
            $y = $l2 ? $l2->val : 0;
            $sum = $carry + $x + $y;
            $carry = intval($sum / 10);
            $curr->next = new ListNode($sum % 10);
            $curr = $curr->next;
            $l1 = $l1 ? $l1->next : null;
            $l2 = $l2 ? $l2->next : null;
        }

        return $dummyHead->next;
    }

}
